File: lib/python/opentoken/src/test/opentoken/io/csv/token_csv_writer_test.py
"""
Copyright (c) Truveta. All rights reserved.
"""

import os
import tempfile

from opentoken.io.csv.token_csv_writer import TokenCSVWriter
from opentoken.processor.token_constants import TokenConstants


class TestTokenCSVWriter:
    """Test cases for TokenCSVWriter."""

    def setup_method(self):
        """Set up test fixtures before each test method."""
        self.temp_file = tempfile.NamedTemporaryFile(mode='w', suffix='.csv', delete=False)
        self.temp_file_path = self.temp_file.name
        self.temp_file.close()
        self.writer = TokenCSVWriter(self.temp_file_path)

    def teardown_method(self):
        """Clean up after each test method."""
        if self.writer:
            self.writer.close()
        if os.path.exists(self.temp_file_path):
            os.unlink(self.temp_file_path)

    def test_write_single_token(self):
        """Test writing a single token to CSV."""
        data = {
            TokenConstants.RULE_ID: "T1",
            TokenConstants.TOKEN: "abc123token",
            TokenConstants.RECORD_ID: "rec-001"
        }

        self.writer.write_token(data)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RuleId,Token,RecordId"

            record = lines[1].strip()
            assert record == "T1,abc123token,rec-001"

    def test_write_multiple_tokens(self):
        """Test writing multiple tokens to CSV."""
        data1 = {
            TokenConstants.RULE_ID: "T1",
            TokenConstants.TOKEN: "token1",
            TokenConstants.RECORD_ID: "rec-001"
        }

        data2 = {
            TokenConstants.RULE_ID: "T2",
            TokenConstants.TOKEN: "token2",
            TokenConstants.RECORD_ID: "rec-002"
        }

        self.writer.write_token(data1)
        self.writer.write_token(data2)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RuleId,Token,RecordId"

            record1 = lines[1].strip()
            assert record1 == "T1,token1,rec-001"

            record2 = lines[2].strip()
            assert record2 == "T2,token2,rec-002"

    def test_header_written_only_once(self):
        """Test that header is written only once."""
        data1 = {
            TokenConstants.RULE_ID: "T1",
            TokenConstants.TOKEN: "token1",
            TokenConstants.RECORD_ID: "rec-001"
        }

        data2 = {
            TokenConstants.RULE_ID: "T2",
            TokenConstants.TOKEN: "token2",
            TokenConstants.RECORD_ID: "rec-002"
        }

        self.writer.write_token(data1)
        self.writer.write_token(data2)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RuleId,Token,RecordId"

            record1 = lines[1].strip()
            assert record1 == "T1,token1,rec-001"

            record2 = lines[2].strip()
            assert record2 == "T2,token2,rec-002"

            # Ensure there are only 3 lines (header + 2 records)
            assert len(lines) == 3

    def test_write_token_with_blank_value(self):
        """Test writing a token with blank value."""
        data = {
            TokenConstants.RULE_ID: "T1",
            TokenConstants.TOKEN: "",
            TokenConstants.RECORD_ID: "rec-001"
        }

        self.writer.write_token(data)
        self.writer.close()

        with open(self.temp_file_path, 'r', encoding='utf-8') as f:
            lines = f.readlines()
            header = lines[0].strip()
            assert header == "RuleId,Token,RecordId"

            record = lines[1].strip()
            assert record == "T1,,rec-001"

File: lib/java/opentoken/src/test/java/com/truveta/opentoken/io/csv/TokenCSVWriterTest.java
/**
 * Copyright (c) Truveta. All rights reserved.
 */
package com.truveta.opentoken.io.csv;

import static org.junit.jupiter.api.Assertions.assertEquals;
import static org.junit.jupiter.api.Assertions.assertTrue;

import com.truveta.opentoken.processor.TokenConstants;
import java.io.BufferedReader;
import java.io.File;
import java.io.FileReader;
import java.nio.file.Files;
import java.util.LinkedHashMap;
import java.util.Map;
import org.junit.jupiter.api.AfterEach;
import org.junit.jupiter.api.BeforeEach;
import org.junit.jupiter.api.Test;

class TokenCSVWriterTest {

    private File tempFile;
    private TokenCSVWriter writer;

    @BeforeEach
    void setUp() throws Exception {
        tempFile = Files.createTempFile("test_token_", ".csv").toFile();
        writer = new TokenCSVWriter(tempFile.getAbsolutePath());
    }

    @AfterEach
    void tearDown() throws Exception {
        if (writer != null) {
            writer.close();
        }
        if (tempFile.exists()) {
            tempFile.delete();
        }
    }

    @Test
    void testWriteSingleToken() throws Exception {
        Map<String, String> data = new LinkedHashMap<>();
        data.put(TokenConstants.RULE_ID, "T1");
        data.put(TokenConstants.TOKEN, "abc123token");
        data.put(TokenConstants.RECORD_ID, "rec-001");

        writer.writeToken(data);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RuleId,Token,RecordId", header);

            String record = reader.readLine();
            assertEquals("T1,abc123token,rec-001", record);
        }
    }

    @Test
    void testWriteMultipleTokens() throws Exception {
        Map<String, String> data1 = new LinkedHashMap<>();
        data1.put(TokenConstants.RULE_ID, "T1");
        data1.put(TokenConstants.TOKEN, "token1");
        data1.put(TokenConstants.RECORD_ID, "rec-001");

        Map<String, String> data2 = new LinkedHashMap<>();
        data2.put(TokenConstants.RULE_ID, "T2");
        data2.put(TokenConstants.TOKEN, "token2");
        data2.put(TokenConstants.RECORD_ID, "rec-002");

        writer.writeToken(data1);
        writer.writeToken(data2);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RuleId,Token,RecordId", header);

            String record1 = reader.readLine();
            assertEquals("T1,token1,rec-001", record1);

            String record2 = reader.readLine();
            assertEquals("T2,token2,rec-002", record2);
        }
    }

    @Test
    void testHeaderWrittenOnlyOnce() throws Exception {
        Map<String, String> data1 = new LinkedHashMap<>();
        data1.put(TokenConstants.RULE_ID, "T1");
        data1.put(TokenConstants.TOKEN, "token1");
        data1.put(TokenConstants.RECORD_ID, "rec-001");

        Map<String, String> data2 = new LinkedHashMap<>();
        data2.put(TokenConstants.RULE_ID, "T2");
        data2.put(TokenConstants.TOKEN, "token2");
        data2.put(TokenConstants.RECORD_ID, "rec-002");

        writer.writeToken(data1);
        writer.writeToken(data2);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RuleId,Token,RecordId", header);

            String record1 = reader.readLine();
            assertEquals("T1,token1,rec-001", record1);

            String record2 = reader.readLine();
            assertEquals("T2,token2,rec-002", record2);

            // Ensure there are no more lines
            String nextLine = reader.readLine();
            assertTrue(nextLine == null || nextLine.isEmpty());
        }
    }

    @Test
    void testWriteTokenWithBlankValue() throws Exception {
        Map<String, String> data = new LinkedHashMap<>();
        data.put(TokenConstants.RULE_ID, "T1");
        data.put(TokenConstants.TOKEN, "");
        data.put(TokenConstants.RECORD_ID, "rec-001");

        writer.writeToken(data);
        writer.close();

        try (BufferedReader reader = new BufferedReader(new FileReader(tempFile))) {
            String header = reader.readLine();
            assertEquals("RuleId,Token,RecordId", header);

            String record = reader.readLine();
            assertEquals("T1,,rec-001", record);
        }
    }
}
